Data/IntMap/Internal.hs

```diff
@@ -980,6 +985,22 @@ alterF f k m = (<$> f mv) $ \fres ->
     Nothing -> maybe m (const (delete k m)) mv
     Just v' -> insert k v' m
   where mv = lookup k m
+-- TODO(m-renaud): Figure out if this should be marked INLINE or NOINLINE.
+-- It needs to be one or the other or else the specialization rule may not fire.
+{-# NOINLINE [1] alterF #-}
+#if MIN_VERSION_base(4,8,0)
+{-# RULES
+"Identity specialize alterF" forall (f :: Maybe a -> Identity (Maybe a)) k m.
+  alterF f k m = Identity $ alter (coerce f) k m
+  #-}
+#endif
+
+#if MIN_VERSION_base(4,9,0)
+{-# RULES
+"Const specialize alterF" forall (f :: Maybe a -> Const x (Maybe a)) k m.
+  alterF f k m = coerce . f $ lookup k m
```

Comment:
No need for `coerce` here! If you really need `Const . getConst` here, then use it, but maybe give it a comment or a type signature. We want to use `coerce` only when needed to prevent silly eta expansion from adding closures and indirection at runtime.

Reply:
Or, of course, to re-type a whole structure from the outside.

Reply:
Yeah, either `Const . getConst` or `coerce` is needed because `f :: Maybe a -> Const c (Maybe a)` but the function returns `Const c (IntMap a)`. Since the second type parameter for `Const` is phantom I would have thought that would be an appropriate use of `coerce`, but I also just learned about Coercible this morning so I may very well be wrong :stuck_out_tongue:.

I've changed it back to `Const . getConst` but I was wondering,  would `Const . getConst` have runtime overhead or should the optimizer figure out that this is just changing a phantom parameter? It seems to me that `Const . getConst` is equivalent to `coerce id` for `Const`.